parser/calcite/CalciteParserServer.java

```diff
@@ -81,24 +94,42 @@ public void parse(
         StreamObserver<CalciteParserProto.CalciteParserReply> responseObserver) {
 
       String q = request.getQuery();
-      Pair<Integer, SqlParseException> r = calcite(q);
-      if (r.getValue() == null) {
-        int i = r.getKey();
-        responseObserver.onNext(
-            CalciteParserProto.CalciteParserReply.newBuilder()
-                .setSql(q.substring(0, i))
-                .setExtension(q.substring(i))
-                .build());
-      } else {
-        responseObserver.onNext(
-            CalciteParserProto.CalciteParserReply.newBuilder()
-                .setError(r.getValue().getCause().getMessage())
-                .build());
-      }
+      Pair<Integer, SqlParseException> r = calciteParse(q);
+      int i = r.getKey();
+      String e = (r.getValue() == null ? "" : r.getValue().getCause().getMessage());
+      responseObserver.onNext(
+          CalciteParserProto.CalciteParserReply.newBuilder()
+          .setIndex(i)
+          .setError(e)
+          .build());
       responseObserver.onCompleted();
     }
 
-    private static int PosToIndex(String query, SqlParserPos pos) {
+    // calciteParse returns <len(query),null> if Calcite parser
+    // accepts query, or returns <pos,null> if a second parsing
+    // accepts the content to the left of the error position from the
+    // first parsing, or <-1,e> if both parsing failed.
+    private static Pair<Integer, SqlParseException> calciteParse(String query) {
```

Comment:
Normally, in `Java` an `exception` is thrown out instead of returning like `go`. In that way, the `pair` in `JavaFX` is also unnecessary.  What do you think?

Reply:
Do you mean returning multiple values by throwing an exception? I am afraid that the exception is not designed for returning values?